infrastructure/src/main/java/org/corfudb/infrastructure/management/FailureDetector.java

```diff
@@ -89,92 +107,77 @@ public PollReport poll(
+                                   SequencerMetrics sequencerMetrics, ImmutableList<String> layoutUnresponsiveNodes,
+                                   FileSystemStats fileSystemStats) {
+
+        final PollConfig pollConfig = this.pollConfig.orElseGet(() -> PollConfig.builder().build());
+        Consumer<MixedBoundRetry> retrySettings = settings -> {
+            settings.setOverallMaxRetryDuration(pollConfig.getMaxDetectionDuration());
+            settings.setMaxRetryDuration(pollConfig.getMaxSleepBetweenRetries());
+            settings.setBaseDuration(pollConfig.getInitSleepBetweenRetries());
+            settings.setRandomPart(pollConfig.getJitterFactor());
+        };
         List<PollReport> reports = new ArrayList<>();
-        for (int iteration = 0; iteration < failureThreshold; iteration++) {
-            PollReport currReport = pollIteration(
-                    allServers, router, epoch, clusterID, sequencerMetrics, layoutUnresponsiveNodes, fileSystemStats
-            );
-            reports.add(currReport);
-
-            Duration restInterval = networkStretcher.getRestInterval(currReport.getElapsedTime());
-            if (!currReport.getFailedNodes().isEmpty()) {
-                networkStretcher.modifyIterationTimeouts();
-
-                Set<String> allReachableNodes = currReport.getAllReachableNodes();
-                tuneRoutersResponseTimeout(
-                        router, allReachableNodes, networkStretcher.getCurrentPeriod()
+        final long start = System.currentTimeMillis();
+        try {
+            IRetry.build(MixedBoundRetry.class, () -> {
+                PollReport currReport = pollIteration(
+                        allServers, router, epoch, clusterID, sequencerMetrics, layoutUnresponsiveNodes, fileSystemStats
                 );
-            }
-
-            // Sleep for the provided poll interval before starting the next iteration
-            Sleep.sleepUninterruptibly(restInterval);
+                reports.add(currReport);
+                // Finish the poll round if we've reached the desired number of reports or when
+                // the current cluster state is not ready.
+                // Cluster state is not ready unless the overall cluster view is updated.
+                // The resulting report will not be used in the failure detection anyway.
+                // Cut this round short and continue on the next FD iteration.
+                if (!currReport.getClusterState().isReady()) {
+                    log.trace("Cluster state is not ready. Skipping iterations.");
+                    throw new RetryExhaustedException();
+                }
+                if (reports.size() == pollConfig.getMaxPollRounds()) {
+                    log.trace("Collected all {} reports.", reports.size());
+                    throw new RetryExhaustedException();
+                }
```

Comment:
This is a little confusing. Instead of using `RetryExhaustedException` why not just `return` ?

Reply:
To be more precise what's confusing is the signal to indicate that the call was successful is being inflated with the same signal that can indicate failure. For example, all polls succeeded vs non of the polls succeeded.

Reply:
Fixed to make it more clear